### PR TITLE
status: Warn for circular deps in buildConfigs

### PR DESCRIPTION
**Pull request by kargakis**
@deads2k PTAL

### CODE REVIEW

**Comment by deads2k** [comment]
I like the idea.  Create a subgraph and run https://godoc.org/github.com/gonum/graph/topo#CyclesIn

**Comment by kargakis** [comment]
@deads2k updated to use CyclesIn

File: pkg/cmd/cli/describe/projectstatus.go [hunk 1]
```diff
@@ -196,6 +201,34 @@ func hasUnresolvedImageStreamTag(g osgraph.Graph) bool {
 	return false
 }
 
+func hasCircularDependencies(g osgraph.Graph) bool {
+	// Filter out all but ImageStreamTag and BuildConfig nodes
+	var nodeFn osgraph.NodeFunc = func(g osgraph.Interface, n graph.Node) bool {
```
Comment (deads2k) [inline pkg/cmd/cli/describe/projectstatus.go]: This is going to be a common desire:  "build me a subgraph with nodes of these kinds and edges of these kinds".  Create a `NodeFunc` and an `EdgeFunc` that do that in `graph.go`, then I think I like it.

**Comment by kargakis** [comment]
updated once more

File: pkg/api/graph/graph.go [hunk 2]
```diff
@@ -353,10 +353,34 @@ func (g Graph) addEdges(edges []graph.Edge, fn EdgeFunc) {
 // node should be included.
 type NodeFunc func(g Interface, n graph.Node) bool
 
+// NewNodeFunc returns a new NodeFunc accepting the provided kinds of nodes
+func NewNodeFunc(kinds ...string) NodeFunc {
```
Comment (deads2k) [inline pkg/api/graph/graph.go]: `NodesOfKind`?  That makes the subgraph command read: `g.Subgraph(NodesOfKind(kind1, kind2), EdgesOfKind(kindA, kindB))`

File: pkg/api/graph/graph.go [hunk 3]
```diff
@@ -353,10 +353,34 @@ func (g Graph) addEdges(edges []graph.Edge, fn EdgeFunc) {
...
+	return func(g Interface, n graph.Node) bool {
+		for _, kind := range kinds {
```
Comment (deads2k) [inline pkg/api/graph/graph.go]: Make the `util.StringSet` of the allowedKinds above, then run an `allowedKinds.Has(g.Kind(n))`

File: pkg/api/graph/graph.go [hunk 4]
```diff
@@ -353,10 +353,34 @@ func (g Graph) addEdges(edges []graph.Edge, fn EdgeFunc) {
...
+}
+
 // EdgeFunc is passed a new graph, an edge in the current graph, and should mutate
 // the new graph as needed. If true is returned, the existing edge will be added to the graph.
 type EdgeFunc func(g Interface, head, tail graph.Node, edgeKinds util.StringSet) bool
 
+// NewEdgeFunc returns a new EdgeFunc accepting the provided kinds of edges
+func NewEdgeFunc(kinds ...string) EdgeFunc {
+	return func(g Interface, head, tail graph.Node, edgeKinds util.StringSet) bool {
+		for _, kind := range kinds {
```
Comment (deads2k) [inline pkg/api/graph/graph.go]: Make the `util.StringSet` of the allowedKinds above, then run an `allowedKinds.HasAny(edgeKinds.List())`.  Also, can you open an issue indicating that we need to fix either `EdgeFunc` or `Subgraph` to properly handle cases where part of an edge should propogate?

File: pkg/api/graph/graph.go (hunk 3 above)
Comment (kargakis) [inline pkg/api/graph/graph.go]: You mean change `...string` to `util.StringSet`? Then we would have to create the stringSet and then pass it to the function, right? Isn't that more cumbersome?

Comment (deads2k) [inline pkg/api/graph/graph.go]:
```
func NewNodeFunc(kinds ...string) NodeFunc {
	allowedKinds := util.NewStringSet(kinds...)
	return func(g Interface, n graph.Node) bool {
		return allowedKinds.Has(g.Kind(n))
	}
}
```

File: pkg/api/graph/graph.go (hunk 4 above)
Comment (kargakis) [inline pkg/api/graph/graph.go]: I don't really understand the issue, can you do it instead and cc me in it?

Comment (deads2k) [inline pkg/api/graph/graph.go]: > I don't really understand the issue, can you do it instead and cc me in it?

Sure.  I introduced it, not you.

File: pkg/api/graph/graph.go (hunk 2 above)
Comment (kargakis) [inline pkg/api/graph/graph.go]: Updated though I didn't squeeze the functions in g.Subgraph because it wouldn't be that nice

**Comment by kargakis** [comment]
@bparees fyi

**Comment by deads2k** [comment]
lgtm [merge]

**Comment by kargakis** [comment]
racy testclient

[merge] again

**Comment by kargakis** [comment]
now an assets flake

[merge]

**Pull request merged**